nep-X1.mediawiki

```diff
@@ -0,0 +1,67 @@
+<pre>
+  NEP: 31
+  Title: Contract Destroy Guideline
+  Author: Owen Zhang <zhangzhihong@ngd.neo.org>
+  Type: Standard
+  Status: Draft
+  Created: 2024-09-11
+</pre>
+
+==Abstract==
+
+This proposal standardizes the <code>destroy</code> function which is necessary for contract storage deletion and contract disposal from the system.
+
+==Motivation==
+
+Contracts that are not relevant anymore may need to be destroyed. Neo N3 has provided contract destroy mechanism but it hasn't been summarized in a standard. This standard will show details.
+
+==Specification==
+
+Neo N3 has the native <code>ContractManagement</code> contract that is responsible for contract destruction via its <code>destroy</code> method. When <code>destroy</code> is executed, it will delete all storage entries of calling contract. It will block the calling contract address using the native <code>Policy</code> contract to prevent redeployment or calling any methods on the contract from that point forward. Contracts that want to be destroyable MUST implement the <code>destroy</code> method as described below.
+
+===Methods===
+
+====destroy====
+
+<pre>
+{
+  "name": "destroy",
+  "safe": false,
+  "parameters": [],
+  "returntype": "Void"
+}
+</pre>
+This method MUST invoke the <code>destroy</code> method of <code>ContractManagement</code> contract when the contract is destroyed.
+
+<code>destroy</code> method can have multi parameters for any specific purpose since it's an operation only for contract owner.
+
+<code>destroy</code> method SHOULD use the <code>SYSCALL</code> <code>CheckWitness</code> to verify the <code>owner</code>. Otherwise anyone will be able to destroy your contract.
+
+Contract SHOULD execute all codes before calling ContractManagement.Destroy() to ensure they can be executed correctly. Especially these contract holding assets, they SHOULD transfer tokens before calling <code>ContractManagement.Destroy()</code> or else the assets will be lost.
```

Comment:
Sometimes it's `ContractManagement.Destroy()` and sometimes it's `<code>destroy</code> method of <code>ContractManagement</code>`. We need some uniformity here (following ABI, of course).